File: 백준/Gold/14676. 영우는 사기꾼？/영우는 사기꾼？.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        Map<Integer, Set<Integer>> info = new HashMap<>();
        int[] pre = new int[n + 1];

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            info.putIfAbsent(a, new HashSet<>());
            info.get(a).add(b);
            pre[b]++;
        }

        Map<Integer, Integer> mine = new HashMap<>();
        String result = "King-God-Emperor";

        for (int i = 0; i < k; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            if (a == 1) {
                if (pre[b] == 0) {
                    mine.putIfAbsent(b, 0);
                    if (mine.get(b) == 0) {
                        if (info.containsKey(b)) {
                            for (int p : info.get(b)) {
                                pre[p]--;
                            }
                        }
                    }
                    mine.put(b, mine.get(b) + 1);
                } else {
                    result = "Lier!";
                }
            } else {
                if (mine.getOrDefault(b, 0) > 0) {
                    mine.put(b, mine.get(b) - 1);
                    if (mine.get(b) == 0) {
                        if (info.containsKey(b)) {
                            for (int p : info.get(b)) {
                                pre[p]++;
                            }
                        }
                    }
                } else {
                    result = "Lier!";
                }
            }
        }

        System.out.println(result);
    }
}

File: 백준/Gold/14676. 영우는 사기꾼？/영우는 사기꾼？.py
import sys
input = lambda : sys.stdin.readline().rstrip()
from collections import defaultdict, Counter
n, m, k = map(int,input().split())
info = defaultdict(set)
pre = [0]*(n+1)
for i in range(m):
    a,b = map(int,input().split())
    info[a].add(b)
    pre[b]+=1
mine = Counter()
result = "King-God-Emperor"
for _ in range(k):
    a,b = map(int,input().split())
    if a==1:
        if not pre[b]:
            if mine[b]==0:
                if b in info:
                    for p in info[b]:
                        pre[p]-=1
            mine[b]+=1
        else:
            result = "Lier!"
    else:
        if mine[b]>0:
            mine[b]-=1
            if mine[b] ==0:
                if b in info:
                    for p in info[b]:
                        pre[p]+=1
        else:
            result = "Lier!"
print(result)